### PR TITLE
Analysis

### PR DESCRIPTION
**Pull request by Gabe-DiAntonio**
Here is my analysis code/tests.

### CODE REVIEW

File: src/Diffusion.jl [hunk 1]
```diff
@@ -108,4 +112,94 @@ function nmolabsorb(σ, steps, pos, r, molecules) #simulates multiple molecules
...
+    if boundary == "absorb"
+        trial = nmolabsorb(σ, steps, pos, r, molecules)
+    elseif boundary == "reflect"
+        trial = nmolreflect(σ, steps, pos, r, molecules)
+    else
+        throw(DomainError("Please specify boudary condition as 'absorb' or 'reflect'"))
+    end
+    lengths = molpathlengths(trial)
+    return percentremaining(lengths)
+    end
```
Comment (timholy) [inline src/Diffusion.jl]: Delete/fix the indentation here and below. While it's not important for the code itself, for humans it makes your code harder to read when it isn't following customary indentation rules. You could try https://github.com/domluna/JuliaFormatter.jl if you don't want to do it manually.

File: src/Diffusion.jl [hunk 2]
```diff
@@ -108,4 +112,94 @@ function nmolabsorb(σ, steps, pos, r, molecules) #simulates multiple molecules
...
+    
+    function molpathlengths(trial) #input the molhistories given my nmolreflect or nmolabsorb
+    lengthsofmolpaths = []         #outputs a vector showing the lengths of each molecules' path
+    for x in 1:length(trial)
```
Comment (timholy) [inline src/Diffusion.jl]: Often people use `i` for integer variables. 

But here's a fun fact: this entire function can be replaced by `map(length, trial)`.

File: src/Diffusion.jl [hunk 3]
```diff
@@ -108,4 +112,94 @@ function nmolabsorb(σ, steps, pos, r, molecules) #simulates multiple molecules
...
+        trial = nmolreflect(σ, steps, pos, r, molecules)
+    else
+        throw(DomainError("Please specify boudary condition as 'absorb' or 'reflect'"))
+    end
+    lengths = molpathlengths(trial)
+    return percentremaining(lengths)
+    end
+    
+    function molpathlengths(trial) #input the molhistories given my nmolreflect or nmolabsorb
+    lengthsofmolpaths = []         #outputs a vector showing the lengths of each molecules' path
```
Comment (timholy) [inline src/Diffusion.jl]:
```suggestion
    lengthsofmolpaths = Int[]         #outputs a vector showing the lengths of each molecules' path
```

File: src/Diffusion.jl [hunk 4]
```diff
@@ -108,4 +112,94 @@ function nmolabsorb(σ, steps, pos, r, molecules) #simulates multiple molecules
...
+    for x in 1:length(trial)
+        push!(lengthsofmolpaths, length(trial[x]))
+    end
+    return lengthsofmolpaths
+    end
+    
+    
+    function numberremaining(lengthsofmolpaths) #input a vector containing the length of each molecules path
+        numberremain = []                       #out puts a vector containing the number of molecules remaining each time step
```
Comment (timholy) [inline src/Diffusion.jl]:
```suggestion
        numberremain = Int[]                       #out puts a vector containing the number of molecules remaining each time step
```

**Review by timholy — COMMENTED** [review]
This is looking pretty good. I left a few specific comments that would probably be worth addressing before mering. But the most important issue is this: separate your "analyze" code from your "run simulation" code. For example, currently you re-run the simulation to calculate the variance if you've just calculated the mean. That implies that you can't extract the mean and variance of a given set of trajectories, and so simulation-to-simulation variability will contribute to differences across runs.

The better design is to run the simulation and get the data, and then pass the trajectories as a variable to the analysis code.

File: src/Diffusion.jl (hunk 3 above)
Comment (timholy) [reply src/Diffusion.jl]: This is important for getting good performance if you do any analysis on this vector. When the elements are of type `Any`, Julia can't optimize the code very well.

File: src/Diffusion.jl [hunk 5]
```diff
@@ -108,4 +113,73 @@ function nmolabsorb(σ, steps, pos, r, molecules) #simulates multiple molecules
...
+        push!(lengthsofmolpaths, length(trial[i]))
+    end
+    return lengthsofmolpaths
+end
+    
+    
+function numberremaining(lengthsofmolpaths) #input a vector containing the length of each molecules path
+    numberremain = Int[]                       #out puts a vector containing the number of molecules remaining each time step
+        for x in 1:maximum(lengthsofmolpaths)
+            number = 0.0
```
Comment (timholy) [inline src/Diffusion.jl:135]:
```suggestion
            number = 0
```
so it stays an `Int`.

**Review by timholy — APPROVED** [review]
This is basically fine, merge at will! (Squash-merge recommended.)

**Comment by timholy** [comment]
Just to check, do you need me to merge this or are you wanting me to review your plotting code? Just let me know.

**Comment by Gabe-DiAntonio** [comment]
Dear Tim,

If you have the time, I'd love a review on the plotting code. However, I
was just pushing everything I did up to Git Hub so that if you want to use
it for the reviewers you have access to it, so just a merge would work
great!

Thank you,
Gabe DiAntonio

**Pull request merged**